[CALCITE-7127] Anti-join with WHERE NOT EXISTS syntax has corrupted condition

dssysolyatin opened the pull request.

mihaibudiu commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -2210,7 +2210,7 @@ private static String toSql(RelNode root, SqlDialect dialect,
         + "GROUP BY \"product\".\"product_id\"\n"
         + "HAVING COUNT(*) > 1) AS \"t2\"\n"
-        + "WHERE \"t2\".\"product_id\" > 100";
+        + "WHERE \"product_id\" > 100";
```
is this output supposed to be affected? It does not match the patterns described in the issue

dssysolyatin replied:
It’s a confusing, but I’ll try to explain it. It is kind of similar. 

# Without fix
If you set a breakpoint inside the `visit(Filter filter)` method at this line:
https://github.com/apache/calcite/blob/cfaadedd8fd34c480395f9e843a73bf4bb8fe459/core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java#L486

...and then inspect `x.aliases` (where x is the SqlImplementor.Result for the Project node), you'll see it contains both `product` and `sales_fact_1997` (Join transfers aliases to SqlImplementor.Result for Project). This is incorrect behavior because the node above the Project should not have access to these aliases. (At least from my point of view)

Next, a new builder is created to transform the `Filter.condition`: (note: At this point we already set mapping between correlation variable and context)
https://github.com/apache/calcite/blob/cfaadedd8fd34c480395f9e843a73bf4bb8fe459/core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java#L487

When this new builder is created, it also creates a new context `builder.context`. This context's `qualified` flag is set to `true` because the previous `Result` `x`  had multiple aliases (`aliases.length` > 1). Because it `t2` is present
https://github.com/apache/calcite/blob/cfaadedd8fd34c480395f9e843a73bf4bb8fe459/core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java#L1902-L1903

Interestingly, the new builder.context itself is valid and correctly contains only the alias t2. This raises a good question: why can't this same context be used for the correlation variable as well? Perhaps I'm misunderstanding something.

# With fix
With the fix, `x.aliases` correctly contains only the `t2` alias. Because there is now only one alias, the `qualified` flag for the new `builder.context` is set to false. This is why the `t2` prefix is no used for product_id.

However, it's important to note that the new context itself is still correct (`builder.context`). It contains also one alias = `t2`

suibianwanwank reviewed (approved):
LGTM, I'm curious, if that's the case, would a type like Project also need similar handling such as `result(builder.select, builder.clauses, e, null)`?

dssysolyatin commented:
@mihaibudiu @suibianwanwank Please rereview PR

Although I still believe the previous fix was correct in theory, but `RelToSqlConverter` does some things that, in my opinion, it shouldn’t and such things should be done by optimizer.

One of such things, that RelToSqlConverter tries to “merge projections” and it is done using alias propogation which I restricted by my previous fix. However, since RelToSqlConverter already has this behavior and some users rely on it, I added a workaround instead. I don’t consider this a proper fix, but at least it doesn’t break anything.

Some example
If I pass the following plan to `RelToSqlConverter`:

```
LogicalProject(EMPNO=[$0])
  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7])
    LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7])
      LogicalTableScan(table=[[scott, EMP]])
```

It generates the following SQL:
```
"SELECT \"EMP\".\"EMPNO\"\n"
    + "FROM \"scott\".\"EMP\" AS \"EMP\""
```
With my previous fix, it produced: 
```
"SELECT \”t0\”.\”EMPNO\"\n"
    + "FROM \"scott\".\"EMP\" AS \"EMP\""
```
But t0 doesn’t exist — which is incorrect.
I’ve also added a couple of tests for these cases.

mihaibudiu commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -553,6 +554,83 @@ private static String toSql(RelNode root, SqlDialect dialect,
   }
 
+  /**
+   * Tests that an identity project is pruned over sca during SQL conversion,
```
typo for "scan"?

mihaibudiu commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -553,6 +554,83 @@ private static String toSql(RelNode root, SqlDialect dialect,
+   */
+  @Test void testPruneIdentityProjectOverScan() {
+    final RelBuilder relBuilder =
+        RelBuilder.create(// Disable merge-project optimization by RelBuilder to generate a plan
```
I would start the comment on the next line
